File: src/controller.rs
/*
 * src/controller/mod.rs
 * tasinput2 - Plugin for creating TAS inputs
 *
 * This file is part of tasinput2.
 *
 * tasinput2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * tasinput2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with tasinput2.  If not, see <https://www.gnu.org/licenses/>.
 */

use crate::Inputs;
use qt_widgets::{
    cpp_core::{CppBox, MutPtr},
    qt_core::{QString, Slot},
    QCheckBox, QHBoxLayout, QLabel, QSpinBox, QVBoxLayout, QWidget,
};
use std::{
    convert::TryInto,
    sync::{Arc, Mutex},
};

type Checkbox = MutPtr<QCheckBox>;
type Spinbox = MutPtr<QSpinBox>;

/// Represents a window used to control inputs.
#[allow(dead_code)]
pub struct Controller<'a> {
    base_window: CppBox<QWidget>,
    inputs: &'a Arc<Mutex<Inputs>>,

    a: Checkbox,
    b: Checkbox,
    z: Checkbox,
    l: Checkbox,
    r: Checkbox,
    start: Checkbox,
    c_up: Checkbox,
    c_left: Checkbox,
    c_down: Checkbox,
    c_right: Checkbox,
    d_up: Checkbox,
    d_left: Checkbox,
    d_down: Checkbox,
    d_right: Checkbox,
    x: Spinbox,
    y: Spinbox,

    a_clicked: Slot<'a>,
    b_clicked: Slot<'a>,
    z_clicked: Slot<'a>,
    l_clicked: Slot<'a>,
    r_clicked: Slot<'a>,
    start_clicked: Slot<'a>,
    c_up_clicked: Slot<'a>,
    c_left_clicked: Slot<'a>,
    c_down_clicked: Slot<'a>,
    c_right_clicked: Slot<'a>,
    d_up_clicked: Slot<'a>,
    d_left_clicked: Slot<'a>,
    d_down_clicked: Slot<'a>,
    d_right_clicked: Slot<'a>,
    x_changed: Slot<'a>,
    y_changed: Slot<'a>,
}

impl<'a> Controller<'a> {
    /// Instantiate a new controller
    pub fn new(input_reference: &'a Arc<Mutex<Inputs>>) -> Controller<'a> {
        let mut base_window = unsafe { QWidget::new_0a() };
        unsafe { base_window.set_window_title(&QString::from_std_str("TAS Input")) };
        let mut layout = unsafe { QVBoxLayout::new_1a(&mut base_window).into_ptr() };

        // macro for creating a checkbox
        macro_rules! checkbox {
            ($name: expr) => {
                unsafe {
                    let mut cbox = QCheckBox::from_q_string(&QString::from_std_str($name));
                    layout.add_widget(&mut cbox);
                    cbox.into_ptr()
                }
            };
        };

        let a = checkbox!("A");
        let b = checkbox!("B");
        let z = checkbox!("Z");
        let l = checkbox!("L");
        let r = checkbox!("R");
        let start = checkbox!("Start");
        let c_up = checkbox!("C Up");
        let c_left = checkbox!("C Left");
        let c_down = checkbox!("C Down");
        let c_right = checkbox!("C Right");
        let d_up = checkbox!("D Up");
        let d_left = checkbox!("D Left");
        let d_down = checkbox!("D Down");
        let d_right = checkbox!("D Right");

        // macro for creating a spin box
        macro_rules! spinbox {
            ($name: expr) => {
                unsafe {
                    /*let mut container = QWidget::new_0a();
                    let mut container_layout = QHBoxLayout::new_0a();

                    // label and spin box
                    let mut spin_label = QLabel::from_q_string(&QString::from_std_str($name));
                    let mut spin_box = QSpinBox::new_0a();
                    spin_box.set_minimum(-127);
                    spin_box.set_maximum(127);

                    container_layout.add_widget(&mut spin_box);
                    container_layout.add_widget(&mut spin_label);

                    container.set_layout(container_layout.into_ptr());

                    layout.add_widget(&mut container);
                    spin_box.into_ptr()*/

                    let mut spin_box = QSpinBox::new_0a();
                    layout.add_widget(&mut spin_box);
                    spin_box.into_ptr()
                }
            };
        }

        let x = spinbox!("X");
        let y = spinbox!("Y");

        // macro for creating a slot that corresponds to a checkbox
        macro_rules! clicked_handler {
            ($cbox: ident) => {
                unsafe {
                    Slot::new(move || {
                        (*input_reference.lock().unwrap()).$cbox = $cbox.is_checked();
                    })
                }
            };
            ($cbox: ident, $dname: ident.$pname: ident) => {
                unsafe {
                    Slot::new(move || {
                        (*input_reference.lock().unwrap()).$dname.$pname = $cbox.is_checked();
                    })
                }
            };
        };

        let a_clicked = clicked_handler!(a);
        let b_clicked = clicked_handler!(b);
        let z_clicked = clicked_handler!(z);
        let l_clicked = clicked_handler!(l);
        let r_clicked = clicked_handler!(r);
        let start_clicked = clicked_handler!(start);
        let c_up_clicked = clicked_handler!(c_up, c.up);
        let c_left_clicked = clicked_handler!(c_left, c.left);
        let c_down_clicked = clicked_handler!(c_down, c.down);
        let c_right_clicked = clicked_handler!(c_right, c.right);
        let d_up_clicked = clicked_handler!(d_up, d.up);
        let d_left_clicked = clicked_handler!(d_left, d.left);
        let d_down_clicked = clicked_handler!(d_down, d.down);
        let d_right_clicked = clicked_handler!(d_right, d.right);

        // macro for creating a slot that corresponds to a spinbox
        macro_rules! changed_handler {
            ($field_name: ident) => {
                unsafe {
                    Slot::new(move || {
                        (*input_reference.lock().unwrap()).$field_name =
                            $field_name.value().try_into().unwrap();
                    })
                }
            };
        };

        let x_changed = changed_handler!(x);
        let y_changed = changed_handler!(y);

        unsafe { base_window.show() };

        let controller = Controller {
            base_window,
            inputs: input_reference,

            a,
            b,
            z,
            l,
            r,
            start,
            c_up,
            c_left,
            c_down,
            c_right,
            d_up,
            d_left,
            d_down,
            d_right,
            x,
            y,

            a_clicked,
            b_clicked,
            z_clicked,
            l_clicked,
            r_clicked,
            start_clicked,
            c_up_clicked,
            c_left_clicked,
            c_down_clicked,
            c_right_clicked,
            d_up_clicked,
            d_down_clicked,
            d_left_clicked,
            d_right_clicked,
            x_changed,
            y_changed,
        };

        unsafe {
            a.clicked().connect(&controller.a_clicked);
            b.clicked().connect(&controller.b_clicked);
            z.clicked().connect(&controller.z_clicked);
            l.clicked().connect(&controller.l_clicked);
            r.clicked().connect(&controller.r_clicked);
            start.clicked().connect(&controller.start_clicked);
            c_up.clicked().connect(&controller.c_up_clicked);
            c_left.clicked().connect(&controller.c_left_clicked);
            c_down.clicked().connect(&controller.c_down_clicked);
            c_right.clicked().connect(&controller.c_right_clicked);
            d_up.clicked().connect(&controller.d_up_clicked);
            d_left.clicked().connect(&controller.d_left_clicked);
            d_down.clicked().connect(&controller.d_down_clicked);
            d_right.clicked().connect(&controller.d_right_clicked);
            dprintln!("Spin boxes");
            x.value_changed().connect(&controller.x_changed);
            y.value_changed().connect(&controller.y_changed);
            dprintln!("Spin boxes finished");
        };

        dprintln!("Sockets connected");

        controller
    }
}

File: src/controller.cpp
/*
 * src/controller.cpp
 * tasinput2 - Plugin for creating TAS inputs
 *
 * This file is part of tasinput2.
 *
 * tasinput2 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * tasinput2 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with tasinput2.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "controller.hpp"

#include "plugin.h"
#include "tasinput2.h"

#include <sstream>
#include <string>
#include <tuple>

#include <QHBoxLayout>
#include <QLabel>
#include <QVBoxLayout>

/* Directionals */

inline QString ty_format(char ty_indicator, const char *after) {
  std::stringstream ss;
  ss << ty_indicator << after;
  return QString::fromStdString(ss.str());
}

DirectionalPanel::DirectionalPanel(QWidget *parent, DirectionalType ty,
                                   Inputs *inputs)
    : QWidget(parent) {
  this->inputs = inputs;
  this->ty = ty;

  // figure out which char represents the type
  char ty_indicator = ty == DT_C ? 'C' : 'D';

  // up cbox
  QVBoxLayout *layout = new QVBoxLayout(this);
  this->up = new QCheckBox(ty_format(ty_indicator, " Up"));
  layout->addWidget(this->up);

  // left/right cbox
  QWidget *middle = new QWidget();
  layout->addWidget(middle);
  QHBoxLayout *middle_layout = new QHBoxLayout(middle);
  this->left = new QCheckBox(ty_format(ty_indicator, " Left"));
  this->right = new QCheckBox(ty_format(ty_indicator, " Right"));

  middle_layout->addWidget(this->left);
  middle_layout->addWidget(this->right);

  // down cbox
  this->down = new QCheckBox(ty_format(ty_indicator, " Down"));
  layout->addWidget(this->down);

  // connect slots
  connect(this->up, SIGNAL(clicked()), this, SLOT(up_checked()));
  connect(this->down, SIGNAL(clicked()), this, SLOT(down_checked()));
  connect(this->left, SIGNAL(clicked()), this, SLOT(left_checked()));
  connect(this->right, SIGNAL(clicked()), this, SLOT(right_checked()));
}

Directional *DirectionalPanel::get_directional() {
  return this->ty == DT_C ? &this->inputs->c : &this->inputs->d;
}

// click handlers
void DirectionalPanel::up_checked() {
  lock_mutex();
  this->get_directional()->up = this->up->isChecked();
  unlock_mutex();
}

void DirectionalPanel::down_checked() {
  lock_mutex();
  this->get_directional()->down = this->down->isChecked();
  unlock_mutex();
}

void DirectionalPanel::left_checked() {
  lock_mutex();
  this->get_directional()->left = this->left->isChecked();
  unlock_mutex();
}

void DirectionalPanel::right_checked() {
  lock_mutex();
  this->get_directional()->right = this->right->isChecked();
  unlock_mutex();
}

/* Button Panel */

ButtonPanel::ButtonPanel(QWidget *parent, Inputs *inputs) : QWidget(parent) {
  this->inputs = inputs;

  QHBoxLayout *layout = new QHBoxLayout(this);

  this->d = new DirectionalPanel(this, DT_D, inputs);
  layout->addWidget(d);

  this->l = new QCheckBox("L");
  layout->addWidget(this->l);

  // Add some buttons
  QWidget *middle = new QWidget();
  layout->addWidget(middle);
  QVBoxLayout *middle_layout = new QVBoxLayout(middle);

  this->a = new QCheckBox("A");
  this->b = new QCheckBox("B");
  this->z = new QCheckBox("Z");
  this->start = new QCheckBox("Start");

  middle_layout->addWidget(this->z);
  middle_layout->addWidget(this->a);
  middle_layout->addWidget(this->b);
  middle_layout->addWidget(this->start);

  // Add R button and C directional
  this->r = new QCheckBox("R");
  layout->addWidget(this->r);

  this->c = new DirectionalPanel(this, DT_C, inputs);
  layout->addWidget(this->c);

  // connect up clicked signals
  connect(this->a, SIGNAL(clicked()), this, SLOT(a_checked()));
  connect(this->b, SIGNAL(clicked()), this, SLOT(b_checked()));
  connect(this->z, SIGNAL(clicked()), this, SLOT(z_checked()));
  connect(this->l, SIGNAL(clicked()), this, SLOT(l_checked()));
  connect(this->r, SIGNAL(clicked()), this, SLOT(r_checked()));
  connect(this->start, SIGNAL(clicked()), this, SLOT(start_checked()));
}

// Button slots
void ButtonPanel::a_checked() {
  lock_mutex();
  this->inputs->a = this->a->isChecked();
  unlock_mutex();
}

void ButtonPanel::b_checked() {
  lock_mutex();
  this->inputs->b = this->b->isChecked();
  unlock_mutex();
}

void ButtonPanel::z_checked() {
  lock_mutex();
  this->inputs->z = this->z->isChecked();
  unlock_mutex();
}

void ButtonPanel::l_checked() {
  lock_mutex();
  this->inputs->l = this->l->isChecked();
  unlock_mutex();
}

void ButtonPanel::r_checked() {
  lock_mutex();
  this->inputs->r = this->r->isChecked();
  unlock_mutex();
}

void ButtonPanel::start_checked() {
  lock_mutex();
  this->inputs->start = this->start->isChecked();
  unlock_mutex();
}

// Helper function to create a labeled spin box.
std::tuple<QWidget *, QSpinBox *> labeled_spinbox(const char *label, int min,
                                                  int max) {
  // general container
  QWidget *container = new QWidget();
  QHBoxLayout *layout = new QHBoxLayout(container);

  // spinbox
  QSpinBox *spinbox = new QSpinBox();
  spinbox->setRange(min, max);
  layout->addWidget(spinbox);

  // label
  QLabel *text = new QLabel(label);
  layout->addWidget(text);

  return std::tuple(container, spinbox);
}

JoystickPanel::JoystickPanel(QWidget *panel, Inputs *inputs) {
  this->inputs = inputs;

  QHBoxLayout *layout = new QHBoxLayout(this);

  // x component
  std::tuple<QWidget *, QSpinBox *> x = labeled_spinbox("X", -127, 127);
  this->x = std::get<1>(x);
  layout->addWidget(std::get<0>(x));

  // y component
  std::tuple<QWidget *, QSpinBox *> y = labeled_spinbox("Y", -127, 127);
  this->y = std::get<1>(y);
  layout->addWidget(std::get<0>(y));

  // signals
  connect(this->x, qOverload<int>(&QSpinBox::valueChanged), this,
          &JoystickPanel::x_changed);
  connect(this->y, qOverload<int>(&QSpinBox::valueChanged), this,
          &JoystickPanel::y_changed);
}

// signals
void JoystickPanel::x_changed(int val) {
  lock_mutex();
  this->inputs->x = val;
  unlock_mutex();
}

void JoystickPanel::y_changed(int val) {
  lock_mutex();
  this->inputs->y = val;
  unlock_mutex();
}

Controller::Controller(QWidget *parent) : QWidget(parent) {
  this->inputs = new Inputs();

  QVBoxLayout *layout = new QVBoxLayout(this);

  this->js = new JoystickPanel(this, this->inputs);
  this->bs = new ButtonPanel(this, this->inputs);

  layout->addWidget(this->js);
  layout->addWidget(this->bs);
}

Controller::~Controller() { delete this->inputs; }

#include "moc_controller.cpp"
